File: android_fragment/commons-core/src/main/java/com/common/core/widget/ActionBarContainer.java
package com.common.core.widget;

import android.content.Context;
import android.util.AttributeSet;
import android.view.View;
import android.widget.Button;
import android.widget.FrameLayout;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.common.core.R;

/**
 * Created by liulun on 2016/11/14.
 * actionBar封装类
 */
public class ActionBarContainer extends FrameLayout {
  private TextView mTitleTextView;
  private FrameLayout mLeftContainer;
  private ImageView mLeftImageView;
  private Button mLeftButton;
  private LinearLayout mRightContainer;
  private ImageButton mRightImageBtn;
  private ImageView mRightImageView;
  private TextView mRightBtn;

  private int mHoldPaddingTop;

  public ActionBarContainer(Context context) {
    this(context, null);
  }

  public ActionBarContainer(Context context, AttributeSet attrs) {
    this(context, attrs, 0);
  }

  public ActionBarContainer(Context context, AttributeSet attrs, int defStyleAttr) {
    super(context, attrs, defStyleAttr);
    init();
  }

  private void init() {
    View.inflate(getContext(), R.layout.view_actionbar_container, this);
    mTitleTextView = (TextView) findViewById(R.id.title_textview);
    mLeftButton = (Button) findViewById(R.id.left_btn);
    mLeftContainer = (FrameLayout) findViewById(R.id.left_container_layout);
    mLeftImageView = (ImageView) findViewById(R.id.left_imageview);
    mRightContainer = (LinearLayout) findViewById(R.id.right_container_layout);
    mRightImageBtn = (ImageButton) findViewById(R.id.right_image_btn);
    mRightImageView = (ImageView) findViewById(R.id.right_imageview);
    mRightBtn = (TextView) findViewById(R.id.right_btn);
  }

  public void setPaddingTop(int paddingTop) {
    mHoldPaddingTop = paddingTop;
    getChildAt(0).setPadding(0, paddingTop, 0, 0);
    getChildAt(0).getLayoutParams().height += paddingTop;
  }

  public void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
    int heightMode = View.MeasureSpec.getMode(heightMeasureSpec);
    if (heightMode == View.MeasureSpec.EXACTLY && mHoldPaddingTop > 0) {
      int heightSize = View.MeasureSpec.getSize(heightMeasureSpec);
      heightMeasureSpec = View.MeasureSpec.makeMeasureSpec(heightSize + mHoldPaddingTop, View.MeasureSpec.EXACTLY);
    }
    super.onMeasure(widthMeasureSpec, heightMeasureSpec);
  }

  public void setTitle(CharSequence title) {
    mTitleTextView.setText(title);
  }

  public void setLeftImage(int resId, OnClickListener onClickListener) {
    showLeftView(mLeftImageView);
    mLeftImageView.setImageResource(resId);
    mLeftContainer.setOnClickListener(onClickListener);
  }

  public void setLeftOnClickListener(OnClickListener onClickListener) {
    showLeftView(mLeftImageView);
    mLeftContainer.setOnClickListener(onClickListener);
  }

  public void setLeftButton(String text, OnClickListener onClickListener) {
    showLeftView(mLeftButton);
    mLeftButton.setText(text);
    mLeftContainer.setOnClickListener(onClickListener);
    mLeftButton.setOnClickListener(onClickListener);
  }

  public void setRightImageBtn(int resId, OnClickListener onClickListener) {
    showRightView(mRightImageBtn);
    mRightImageBtn.setImageResource(resId);
    mRightImageBtn.setOnClickListener(onClickListener);
    mRightContainer.setOnClickListener(onClickListener);
  }

  public void setRightImageView(int resId, OnClickListener onClickListener) {
    showRightView(mRightImageView);
    mRightImageView.setImageResource(resId);
    mRightImageView.setOnClickListener(onClickListener);
    mRightContainer.setOnClickListener(onClickListener);
  }

  public void setLeftVisible(boolean visible) {
    mLeftContainer.setVisibility(visible ? VISIBLE : GONE);
  }

  public boolean isLeftVisible() {
    return mLeftContainer.getVisibility() == VISIBLE;
  }

  public void setRightVisible(boolean visible) {
    mRightContainer.setVisibility(visible ? VISIBLE : GONE);
  }

  /**
   * 按钮设置
   *
   * @param text
   */
  public void setRightBtn(String text) {
    setRightBtn(text, null);
  }

  public void setRightBtn(String text, OnClickListener onClickListener) {
    showRightView(mRightBtn);
    mRightBtn.setText(text);
    mRightBtn.setOnClickListener(onClickListener);
    mRightContainer.setOnClickListener(onClickListener);
  }

  private void showRightView(View rightView) {
    mRightContainer.setVisibility(VISIBLE);
    for (int i = 0; i < mRightContainer.getChildCount(); i++) {
      mRightContainer.getChildAt(i).setVisibility(GONE);
    }
    rightView.setVisibility(VISIBLE);
  }

  private void showLeftView(View leftView) {
    mLeftContainer.setVisibility(VISIBLE);
    for (int i = 0; i < mLeftContainer.getChildCount(); i++) {
      mLeftContainer.getChildAt(i).setVisibility(GONE);
    }
    leftView.setVisibility(VISIBLE);
  }

}

File: example/commons-core/src/main/java/com/common/core/widget/ActionBarContainer.kt
package com.common.core.widget

import android.content.Context
import android.util.AttributeSet
import android.view.View
import android.widget.Button
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import com.common.core.R
import com.common.core.utils.ScreenUtils

/**
 * Created by liulun on 2016/11/14.
 * actionBar封装类
 */
class ActionBarContainer @JvmOverloads constructor(context: Context, attrs: AttributeSet? = null, defStyleAttr: Int = 0) : FrameLayout(context, attrs, defStyleAttr) {
    private var mTitleTextView: TextView? = null
    private var mLeftContainer: FrameLayout? = null
    private var mLeftImageView: ImageView? = null
    private var mLeftButton: Button? = null
    private var mRightContainer: LinearLayout? = null
    private var mRightImageBtn: ImageButton? = null
    private var mRightImageView: ImageView? = null
    private var mRightBtn: TextView? = null

    private var mHoldPaddingTop: Int = 0

    var isLeftVisible: Boolean
        get() = mLeftContainer!!.visibility == View.VISIBLE
        set(visible) {
            mLeftContainer!!.visibility = if (visible) View.VISIBLE else View.GONE
        }

    init {
        init()
    }

    private fun init() {
        View.inflate(context, R.layout.view_actionbar_container, this)
        mTitleTextView = findViewById(R.id.title_textview)
        mLeftButton = findViewById(R.id.left_btn)
        mLeftContainer = findViewById(R.id.left_container_layout)
        mLeftImageView = findViewById(R.id.left_imageview)
        mRightContainer = findViewById(R.id.right_container_layout)
        mRightImageBtn = findViewById(R.id.right_image_btn)
        mRightImageView = findViewById(R.id.right_imageview)
        mRightBtn = findViewById(R.id.right_btn)
    }

    fun setPaddingTop(paddingTop: Int) {
        mHoldPaddingTop = paddingTop
        getChildAt(0).setPadding(0, paddingTop, 0, 0)
        getChildAt(0).layoutParams.height += paddingTop
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var heightMeasureSpec = heightMeasureSpec
        val heightMode = View.MeasureSpec.getMode(heightMeasureSpec)
        if (heightMode == View.MeasureSpec.EXACTLY && mHoldPaddingTop > 0) {
            val heightSize = View.MeasureSpec.getSize(heightMeasureSpec)
            heightMeasureSpec = View.MeasureSpec.makeMeasureSpec(heightSize + mHoldPaddingTop, View.MeasureSpec.EXACTLY)
        }
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
    }

    fun setTitle(title: CharSequence) {
        mTitleTextView!!.text = title
    }

    fun setLeftImage(resId: Int, onClickListener: View.OnClickListener) {
        showLeftView(mLeftImageView)
        mLeftImageView!!.setImageResource(resId)
        mLeftContainer!!.setOnClickListener(onClickListener)
    }

    fun setLeftOnClickListener(onClickListener: View.OnClickListener) {
        showLeftView(mLeftImageView)
        mLeftContainer!!.setOnClickListener(onClickListener)
    }

    fun setLeftButton(text: String, onClickListener: View.OnClickListener) {
        showLeftView(mLeftButton)
        mLeftButton!!.text = text
        mLeftContainer!!.setOnClickListener(onClickListener)
        mLeftButton!!.setOnClickListener(onClickListener)
    }

    fun setRightImageBtn(resId: Int, onClickListener: View.OnClickListener) {
        showRightView(mRightImageBtn)
        mRightImageBtn!!.setImageResource(resId)
        mRightImageBtn!!.setOnClickListener(onClickListener)
        mRightContainer!!.setOnClickListener(onClickListener)
    }

    fun setRightImageView(resId: Int, onClickListener: View.OnClickListener) {
        showRightView(mRightImageView)
        mRightImageView!!.setImageResource(resId)
        mRightImageView!!.setOnClickListener(onClickListener)
        mRightContainer!!.setOnClickListener(onClickListener)
    }

    fun setRightVisible(visible: Boolean) {
        mRightContainer!!.visibility = if (visible) View.VISIBLE else View.GONE
    }

    @JvmOverloads
    fun setRightBtn(text: String, onClickListener: View.OnClickListener? = null) {
        showRightView(mRightBtn)
        mRightBtn!!.text = text
        mRightBtn!!.setOnClickListener(onClickListener)
        mRightContainer!!.setOnClickListener(onClickListener)
    }

    private fun showRightView(rightView: View?) {
        mRightContainer!!.visibility = View.VISIBLE
        for (i in 0 until mRightContainer!!.childCount) {
            mRightContainer!!.getChildAt(i).visibility = View.GONE
        }
        rightView!!.visibility = View.VISIBLE
    }

    private fun showLeftView(leftView: View?) {
        mLeftContainer!!.visibility = View.VISIBLE
        for (i in 0 until mLeftContainer!!.childCount) {
            mLeftContainer!!.getChildAt(i).visibility = View.GONE
        }
        leftView!!.visibility = View.VISIBLE
    }
}
/**
 * 按钮设置
 *
 * @param text
 */
